go/vt/servenv/embedder.go

```diff
@@ -0,0 +1,444 @@
+		prefix:      prefix,
+		handleFuncs: make(map[string]*handleFunc),
+	}
+	if instanceName != "" {
+		e.sp = newStatusPage(instanceName)
+	}
+	embeds[instanceName] = e
+}
+
+func (e *embedder) resetLocked(prefix string) {
+	e.prefix = prefix
+	for _, hf := range e.handleFuncs {
+		hf.Set(nil)
+	}
+
+	for _, vmap := range globalStatVars {
+		vmap.mu.Lock()
+		delete(vmap.vars, e.name)
+		vmap.mu.Unlock()
+	}
+
+	if e.sp != nil {
+		e.sp.reset()
+	}
+}
+
+// findOrCreateEmbedder returns an existing embedder or
+// creates a new one.
+func findOrCreateEmbedder(instanceName string) *embedder {
+	if e, ok := embeds[instanceName]; ok {
+		return e
+	}
+	e := &embedder{
+		name:        instanceName,
+		prefix:      "",
+		handleFuncs: make(map[string]*handleFunc),
+	}
+	if instanceName != "" {
+		e.sp = newStatusPage(instanceName)
+	}
+	embeds[instanceName] = e
+	return e
+}
+
+// URLPrefix returns the URL prefix for all the embedder.
+func URLPrefix(instanceName string) string {
+	// There are two other places where this logic is duplicated:
+	// status.go and go/vt/vtgate/discovery/healthcheck.go.
+	if instanceName == "" {
+		return ""
+	}
+	return "/" + instanceName
+}
+
+// HandleFunc sets or overwrites the handler for url. If embedder has a name,
+// url remapped from /path to /name/path. If name is empty, the request
+// is passed through to http.HandleFunc.
+func HandleFunc(instanceName, url string, f func(w http.ResponseWriter, r *http.Request)) {
+	if instanceName == "" {
+		http.HandleFunc(url, f)
+		return
+	}
+
+	embedmu.Lock()
+	defer embedmu.Unlock()
+	e := findOrCreateEmbedder(instanceName)
+
+	hf, ok := e.handleFuncs[url]
+	if ok {
+		hf.Set(f)
+		return
+	}
+	hf = &handleFunc{f: f}
+	e.handleFuncs[url] = hf
+
+	http.HandleFunc(URLPrefix(e.name)+url, func(w http.ResponseWriter, r *http.Request) {
+		if f := hf.Get(); f != nil {
+			f(w, r)
+		}
+	})
+}
+
+// AddInstanceStatusPart adds a status part to the status page. If instance has a name,
+// the part is added to a url named /name/debug/status. Otherwise, it's /debug/status.
+func AddInstanceStatusPart(instanceName, banner, frag string, f func() interface{}) {
+	if instanceName == "" {
+		AddStatusPart(banner, frag, f)
+		return
+	}
+
+	embedmu.Lock()
+	defer embedmu.Unlock()
+	e := findOrCreateEmbedder(instanceName)
+	e.sp.addStatusPart(banner, frag, f)
+}
+
+// NewCountersFuncWithMultiLabels creates a name-spaced equivalent for stats.NewCountersFuncWithMultiLabels.
+func NewCountersFuncWithMultiLabels(instanceName, name, help string, labels []string, f func() map[string]int64) *stats.CountersFuncWithMultiLabels {
+	// If instanceName is empty, it's a pass-through.
+	// If name is empty, it's an unexported var.
+	if instanceName == "" || name == "" {
+		return stats.NewCountersFuncWithMultiLabels(name, help, labels, f)
+	}
+
+	embedmu.Lock()
+	defer embedmu.Unlock()
+	e := findOrCreateEmbedder(instanceName)
+
+	if vmap, ok := globalStatVars[name]; ok {
+		vmap.Set(e.name, f)
```

Comment:
I'm confused about this... when two separate embedded instances register the same stat var (albeit with different instanceNames), won't this make it so the last one to register "wins" and overrides the other?

Logically, it seems to me we need to register ONE metric under `name` that then calls a function which collects the values from each instance (tagged with the instanceName) as one big CountersFuncWithMultiLabels.